src/algorithms/pid/ParticlesWithPID.cc

```diff
@@ -0,0 +1,325 @@
+                in_track_p = in_track_p + ( in_track_point.momentum / in_track.points_size() );
+            auto in_track_eta = edm4hep::utils::eta(in_track_p);
+            auto in_track_phi = edm4hep::utils::angleAzimuthal(in_track_p);
+
+            // calculate dist(eta,phi)
+            auto match_dist = std::hypot(
+                    in_part_eta - in_track_eta,
+                    in_part_phi - in_track_phi
+                    );
+
+            // check if the match is close enough: within user-specified tolerances
+            auto match_is_close =
+                std::abs(in_part_eta - in_track_eta) < m_cfg.etaTolerance &&
+                std::abs(in_part_phi - in_track_phi) < m_cfg.phiTolerance;
+            if (match_is_close)
+                prox_match_list.push_back(ProxMatch{match_dist, in_pid_idx});
+
+            // logging
+            m_log->trace("  - (eta,phi) = ( {:>5.4}, {:>5.4} deg ),  match_dist = {:<5.4}{}",
+                    in_track_eta,
+                    in_track_phi * 180.0 / M_PI,
+                    match_dist,
+                    match_is_close ? " => CLOSE!" : ""
+                    );
+
+        } // end loop over input CherenkovParticleID objects
+
+        // check if at least one match was found
+        if (prox_match_list.size() == 0) {
+            m_log->warn("no matching CherenkovParticleID found for this particle");
+            return false;
+        }
+
+        // choose the closest matching CherenkovParticleID object corresponding to this input reconstructed particle
+        auto closest_prox_match = *std::min_element(
+                prox_match_list.begin(),
+                prox_match_list.end(),
+                [] (ProxMatch a, ProxMatch b) { return a.match_dist < b.match_dist; }
+                );
+        auto in_pid_matched = in_pids.at(closest_prox_match.pid_idx);
+        m_log->trace("  => best match: match_dist = {:<5.4} at idx = {}",
+                closest_prox_match.match_dist,
+                closest_prox_match.pid_idx
+                );
+
+        // convert `CherenkovParticleID` object's hypotheses => set of `ParticleID` objects
+        auto out_pid_index_map = ConvertParticleID::ConvertToParticleIDs(in_pid_matched, out_pids, true);
+        if (out_pid_index_map.size() == 0) {
+            m_log->error("found CherenkovParticleID object with no hypotheses");
+            return false;
+        }
+
+        // relate matched ParticleID objects to output particle
+        for (const auto& [out_pids_index, out_pids_id] : out_pid_index_map) {
+            const auto& out_pid = out_pids->at(out_pids_index);
+            if (out_pid.id() != out_pids_id) { // sanity check
+                m_log->error("indexing error in `edm4eic::ParticleID` collection");
+                return false;
+            }
+            in_part.addToParticleIDs(out_pid);
+        }
+        in_part.setParticleIDUsed(in_part.getParticleIDs().front()); // highest likelihood is the first
```

Comment:
```suggestion
        in_part.setParticleIDUsed(in_part.getParticleIDs().at(0)); // highest likelihood is the first
```
`front()` doesn't exist

Reply:
Right. It's a map.

Reply:
Or no, it's a partial implementation of the std::vector interface... Like how operator== isn't defined on the iterators, but operator!= is defined...